File: iosInblo/iosInblo/App/horse_list/HorseListViewModel.swift
//
//  HorseListViewModel.swift
//  iosInblo
//
//  Created by Jhondee Diaz on 10/22/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import shared

class HorseListViewModel: ObservableObject {
    
    @AppStorage("stableId") var stableId: Int = 0
    @AppStorage("userId") var userId: Int = 0
    @AppStorage("horseArchiveMode") var horseArchiveMode: Bool = false
    @Published var alert: AlertContent?

    
    let searchHorses: SearchHorses
    let getUserNames: GetUsernames
    let addHorse: AddHorse
    let archiveHorse: ArchiveHorse
    let restoreArchiveHorse: RestoreArchivedHorse
    
    @Published var horses = [Horse]()
    @Published var users = [UserDto]()
    
    init(
        searchHorses: SearchHorses,
        getUserNames: GetUsernames,
        addHorse: AddHorse,
        archiveHorse: ArchiveHorse,
        restoreArchiveHorse: RestoreArchivedHorse
    ){
        self.searchHorses = searchHorses
        self.getUserNames = getUserNames
        self.addHorse = addHorse
        self.archiveHorse = archiveHorse
        self.restoreArchiveHorse = restoreArchiveHorse

//        loadHorses()
        loadUserNames(stabId: String(stableId))
        
    }
    
    func loadHorses(
        isArchived: Bool = false
    ){
        do {
            searchHorses.invoke(
                offset: 0,
                query: nil,
                isArchived: isArchived,
                stableId: Int32(stableId)
            ).collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                if dataState != nil {
                    
                    if(dataState is DataStateData){
                        
                        let horses = dataState?.value(forKey: "data") as! [Horse]
                        self.horses = horses
                    }
                
                }
            })
        }
    }
    
    func loadUserNames(stabId: String? = nil,
                       exId: String? = nil){
        do {
            getUserNames.invoke(stableId: stabId, excludeId: exId).collectCommon(coroutineScope: nil, callback: { dataState in
                if dataState != nil {
                    if(dataState is DataStateData){
                        print(dataState)
                        let getUserResponse = dataState?.value(forKey: "data") as! GetUsersResponse
                        self.users = getUserResponse.data
                    }
                }
                
            })
        }
    }
    
    func addHorse(
        name: String,
        userId: Int,
        ownerName: String,
        farmName: String,
        trainingFarmName: String,
        sex: String,
        color: String,
        _class: String,
        birthDate: Date,
        father: String,
        mother: String,
        motherFatherName: String,
        totalStake: String,
        notes: String,
        horseId: String? = nil
    ){
        let dateFormatterPrint = DateFormatter()
        dateFormatterPrint.dateFormat = "yyyy-MM-dd"
        
        let formattedBirthDate = dateFormatterPrint.string(from: birthDate)
        let id = horseId == nil ? nil : horseId
//
        addHorse.invoke(name: name, userId: Int32(userId), stableId: Int32(stableId), ownerName: ownerName, farmName: farmName, trainingFarmName: trainingFarmName, sex: sex, color: color, _class: _class, birthDate: formattedBirthDate, father: father, mother: mother, motherFatherName: motherFatherName, totalStake: Double(totalStake) ?? 0.0, notes: notes, horseId: horseId)
            .collectCommon(coroutineScope: nil, callback: { dataState in
                if(dataState is DataStateData){
                    print("data: \(dataState)")
                    let data = dataState?.value(forKey: "data") as! Horse
                    
                    let message: String = {
                        if horseId == nil {
                            return "管理馬の登録が完了しました！"
                        }else{
                            return "管理馬の修正が完了しました！"
                        }
                    }()
                    
                    self.alert = AlertContent(title: "", body: message, type: getAlertTypeByCode(200))
                    self.loadHorses()
                }else{
                    print("error: \(dataState)")
                }
            })
    }
    
    func archiveHorse(
        horseId: String
    ){
        archiveHorse.invoke(horseId: horseId)
            .collectCommon(coroutineScope: nil, callback: { dataState in
                if(dataState is DataStateData){
                    print("data: \(dataState)")
                    let data = dataState?.value(forKey: "data") as! BooleanResponse
                    
                    let message = data.meta.message
                    
                    self.alert = AlertContent(title: "", body: message, type: getAlertTypeByCode(200))
                    self.loadHorses()
                }
            })
    }
    
    func restoreArchiveHorse(
        horseId: String
    ){
        restoreArchiveHorse.invoke(horseId: horseId)
            .collectCommon(coroutineScope: nil, callback: { dataState in
                if(dataState is DataStateData){
                    print("data: \(dataState)")
                    let data = dataState?.value(forKey: "data") as! BooleanResponse
                    
                    let message = data.meta.message
                    
                    self.alert = AlertContent(title: "", body: message, type: getAlertTypeByCode(200))
                    
                    self.loadHorses(isArchived: true)
                }
            })
    }
}

File: androidInblo/src/main/java/com/colinjp/inblo/android/presentation/ui/horse_list/HorseListViewModel.kt
package com.colinjp.inblo.android.presentation.ui.horse_list

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.colinjp.inblo.android.domain.util.UserPreferences
import com.colinjp.inblo.datasource.network.horse_detail.DropdownData
import com.colinjp.inblo.domain.model.Horse
import com.colinjp.inblo.domain.model.User
import com.colinjp.inblo.interactors.horse_detail.GetHorseStables
import com.colinjp.inblo.interactors.horse_list.AddHorse
import com.colinjp.inblo.interactors.horse_list.ArchiveHorse
import com.colinjp.inblo.interactors.horse_list.GetUsernames
import com.colinjp.inblo.interactors.horse_list.SearchHorses
import com.colinjp.inblo.util.DataState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject


@HiltViewModel
class HorseListViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    private val searchHorses: SearchHorses,
    private val addHorseUseCase: AddHorse,
    private val getUsernames: GetUsernames,
    private val getHorseStables: GetHorseStables,
    private val archiveHorse: ArchiveHorse,
    private val userPreferencesFlow: Flow<UserPreferences>
): ViewModel() {

    val sortOrder = MutableStateFlow<SortOrder>(SortOrder.BY_DATE)
    val horses = MutableStateFlow<DataState<List<Horse>>>(DataState.Empty)
    val users = MutableStateFlow<DataState<List<User>>>(DataState.Empty)


    fun getHorses(sortOrder: SortOrder,stableId: Int){
        viewModelScope.launch(IO){
            searchHorses.invoke(stableId = stableId
            ).onEach {
                horses.value = it
            }.launchIn(viewModelScope)
        }
    }

    fun getUsernames(stableId: String? = null,
                     excludeId: String? = null){
        viewModelScope.launch(IO) {
            getUsernames.invoke(
                stableId,excludeId
            ).onEach { userResponse ->
                when(userResponse){
                    is DataState.Data -> {
                        users.value = DataState.Data(userResponse.data.data.map { userDto ->  User.createFromDto(userDto) })
                    }
                    DataState.Empty -> {

                    }
                    is DataState.Error -> {
                        //TODO:

                    }
                    is DataState.Loading -> {
                        //TODO:

                    }
                }
            }.launchIn(viewModelScope)
        }
    }



    // listen to save horse
    private val addEventChannel = Channel<AddEvent>()
    val addEventFlow = addEventChannel.receiveAsFlow()

    private val updateEventChannel = Channel<AddEvent>()
    val updateEventFlow = updateEventChannel.receiveAsFlow()

    sealed class AddEvent {
        data class Success(val message: String): AddEvent()
        data class Error(val errorMessage: String): AddEvent()
        data class Loading(val isLoading: Boolean): AddEvent()
    }

    fun sortHorses(
        sort: SortOrder
    ){
        sortOrder.value = sort
    }

    fun addHorse(
        name: String,
        userId: String,
        stableId: Int,
        ownerName: String,
        farmName: String,
        trainingFarmName: String,
        sex: String,
        color: String,
        _class: String,
        birthDate: String,
        father: String,
        mother: String,
        motherFatherName: String,
        totalStake: Double,
        notes: String,
        horseId: Int?
    ){
        viewModelScope.launch(IO){
            addHorseUseCase.invoke(
                name, userId.toInt(),stableId,ownerName, farmName, trainingFarmName, sex, color, _class, birthDate, father, mother, motherFatherName, totalStake, notes, horseId?.toString()
            ).onEach {
                when(it){
                    is DataState.Data -> {
                        Timber.v(it.toString())
                        if(horseId == null){
                            addEventChannel.send(AddEvent.Success("管理馬の登録が完了しました！"))
                        }else{
                            updateEventChannel.send(AddEvent.Success("管理馬の修正が完了しました！"))
                        }
                        userPreferencesFlow.collect { userPref ->
                            getHorses(SortOrder.BY_DATE,userPref.stableId)
                        }
                    }
                    DataState.Empty -> {

                    }
                    is DataState.Error -> {
                        Timber.e(it.toString())
                        if(horseId == null){
                            addEventChannel.send(AddEvent.Error("通信に問題があります。後ほどアクセスしてください。"))
                        }else{
                            updateEventChannel.send(AddEvent.Success("通信に問題があります。後ほどアクセスしてください。"))
                        }
                    }
                    is DataState.Loading -> {

                    }
                }
            }.launchIn(viewModelScope)
        }
    }

    fun archiveHorse(
        horseId: String
    ){
        viewModelScope.launch(IO){
            archiveHorse.invoke(
                horseId
            ).onEach {
                when(it){
                    is DataState.Data -> {
                        Timber.v(it.toString())
                        if(it.data.meta.code == 200){
                            addEventChannel.send(AddEvent.Success("Horse successfully archived!"))
                        }
                        userPreferencesFlow.collect { userPref ->
                            getHorses(SortOrder.BY_DATE,userPref.stableId)
                        }
                    }
                    DataState.Empty -> {

                    }
                    is DataState.Error -> {
                        Timber.e(it.toString())
                        addEventChannel.send(AddEvent.Error("Something Went wrong."))
                    }
                    is DataState.Loading -> {

                    }
                }
            }.launchIn(viewModelScope)
        }
    }

}

enum class SortOrder { BY_DATE, BY_SCHEDULE, BY_DECIDED  }